Add default config location loading

cthoyt opened the pull request:
Closes #5

Still to-do:

- [x] Make configurable via the _load function

cthoyt committed:
First try at default config location

References scolby/easy_config#5

scolby33 commented:
If `_load(_whatever_argument_name_you_choose)` is unset or None or whatever, don't look for a default in the environment. (In case that wasn't clear from how I said things before.)

cthoyt committed:
Use pre-existing fixture for test config file

cthoyt committed:
Add reference implementation

Also it looks like testing is having an issue with the setting/unsetting of environment variables

cthoyt committed:
Fix wrong testing

cthoyt committed:
Fix scoping of fixtures

cthoyt committed:
Merge branch 'develop' into add-default-config-envvar

cthoyt committed:
Fix vulture

cthoyt committed:
Update docs

scolby33 commented on src/easy_config/__init__.py:
```diff
@@ -148,12 +149,14 @@ def load(
         2. values from files passed in the additional_files parameter, in order
-        3. values from the environment
-        4. values passed as keyword arguments to this method (useful for values specified on the command line)
+        3. values from the file specified by the config file specified by the environment variable _lookup_config_envvar
```
"values from the file specified by the path in the environment variable _lookup_config_envvar" reads better.

Also it doesn't specify the NAME_ and upper transformation here.

scolby33 commented on src/easy_config/__init__.py:
```diff
@@ -165,6 +168,10 @@ def load(
                 values.update(cls._load_file(f))
+        if _lookup_config_envvar is not None:
+            envvar = f'{cls.NAME.upper()}_{_lookup_config_envvar.upper()}'
+            f = os.environ[envvar]
```
What happens if the envvar is unset?

cthoyt replied:
muahaha explosions. Should we make a custom wrapper?

scolby33 replied:
Nah, just use `.get` and check for None or whatever or put a try around it.

Add a test.

scolby33 commented on tests/conftest.py:
```diff
@@ -21,8 +21,19 @@ def example_ini(tmpdir_factory) -> Path:
     """Add example values to the environment."""
     os.environ['MYPROGRAM_NUMBER'] = '4'
     os.environ['MYPROGRAM_FLAG'] = 'True'
+    yield  # cleanup of newenvironment variables is necessary
```
space

scolby33 commented on tests/test_easy_config.py:
```diff
@@ -47,6 +48,18 @@ def test_load_environment(example_env):
+def test_load_from_env(example_config_env):
+    """Test EasyConfig._load_environment."""
+    assert 'MYPROGRAM_FLAG' not in os.environ
+    assert 'MYPROGRAM_NUMBER' not in os.environ
```
Can you assert with a message or add a comment here about why you're doing these assertions?

cthoyt replied:
yeah, but do you think it's better to remove them now that the issue i was checking for has been resolved?

scolby33 replied:
Honestly I'm happy to leave them as a reminder of what happened if it happens again.

scolby33 reviewed (changes requested):
A few minor changes.

cthoyt committed:
Update documentation

[skip ci]

cthoyt committed:
More docs

[skip ci]

cthoyt committed:
Add error handling and testing

scolby33 commented on src/easy_config/__init__.py:
```diff
@@ -156,7 +156,10 @@ def load(
+        :param _lookup_config_envvar: the environment variable that contains the config file location. Like the loading
+         from the environment, this value will be uppercased and post-pended to the program name. For example, the
+         _lookup_config_envvar "config" for an instance with the NAME "myprogram" will result in a search for the
+         environment variable "MYPROGRAM_CONFIG" for the path to the configuration file.
```
perfect

scolby33 commented on src/easy_config/__init__.py:
```diff
@@ -165,6 +171,21 @@ def load(
+            if not f:
+                raise ValueError(f'Empty value set for {envvar} in environment')
+
+            if not os.path.exists(f):
```
Don't need this, configparser transparently handles nonexistent files.

I'll add this to the documentation for overriding `_load_file` to say that it should never fail.

Unless you think this should be an error?

cthoyt replied:
idk, I leave this up to you to decide. If there's invalid configuration for the configuration (meta-configuration?) should it explode? I happen to like programs that explode when it's my fault rather than saying "that's fine, charlie, just go on"

scolby33 replied:
I think the possibility of someone having a default configuration location that doesn't need to exist is the better case to optimize for.

Maybe I'll go back and add some `logger.debug`'s for things like that.

cthoyt replied:
Okay, fair enough!

scolby33 commented on src/easy_config/__init__.py:
```diff
@@ -165,6 +171,21 @@ def load(
+                raise KeyError(f'No value set for {envvar} in environment') from e
+
+            if not f:
+                raise ValueError(f'Empty value set for {envvar} in environment')
```
These two should definitely not be errors but should just be ignored.

scolby33 replied:
(Like, given the situation that a program wants to allow parsing a config file passed in the environment but doesn't require it and doesn't want to use a cli option to turn that on or off.)

cthoyt replied:
Your choice. Maybe you can go ahead and remove the parts you don't want

scolby33 replied:
#10

scolby33 committed:
Remove overzealous erroring.

scolby33 committed:
Allow higher complexity with flake8 and xenon because the load method is getting a bit busy.

cthoyt commented on src/easy_config/__init__.py:
```diff
@@ -140,6 +140,7 @@ def load(
         *,
         _parse_files: bool = True,
         _parse_environment: bool = True,
+        _lookup_config_envvar: Optional[str] = None,
```
Should this be configurable on a class level? What's really the difference if things are configured there or during the call to the function, which in theory only gets called once since this *should* be a singleton?

cthoyt replied:
just a thought. I'm happy with how this works

scolby33 replied:
Yeah I'm happy too.

scolby33 replied:
I mean, happy about this. `_lookup_config_envvar` doesn't make me happy in a general sense.

cthoyt commented on src/easy_config/__init__.py:
```diff
@@ -165,6 +171,11 @@ def load(
                 values.update(cls._load_file(f))
+        if _lookup_config_envvar is not None:
+            envvar = f'{cls.NAME.upper()}_{_lookup_config_envvar.upper()}'
+            f = os.environ.get(envvar)
```
def more simple now

cthoyt commented on tests/test_easy_config.py:
```diff
@@ -47,6 +48,45 @@ def test_load_environment(example_env):
+
+def test_load_from_env_unset():
+    """Test looking up a config file via an environment variable that isn't set."""
+    a = ExampleConfig.load(number=3, floaty_number=5, flag=False, word='hello', _lookup_config_envvar='config')
```
Idk but `_lookup_config_envvar` just looks ugly here

cthoyt replied:
I can get over it

scolby33 commented:
I merged at the CLI and github didn't detect it...

cthoyt commented:
Aww :/
